### ISSUE TITLE
perturb_h has no effect in combination with X2C relativistic corrections

### ISSUE THREAD

**Issue by tovrstra** [psi4/psi4#1758]
Possibly the following could (also) be due to my misunderstanding of how the input is supposed to work. My apologies if that is the case. In the following example, the perturbation with the external field seems to be ignored:

```
molecule {
  0 1
  He 0.0 0.0 0.0
  symmetry c1  # field breaks symmetry
}

set {
  reference   rhf
  basis  aug-cc-pvtz
  relativistic x2c
  perturb_h true
  perturb_with dipole
  perturb_dipole [0.0, 0.0, 0.1]
}

energy('scf')
```

I've attached a series of examples, with non-relativistic, dkh and x2c calculations on a dummy test case: [reltest.zip](https://github.com/psi4/psi4/files/3884841/reltest.zip)

```
grep "Total Energy =" *.out
dkh_0.out:    Total Energy =                         -2.8613053380694344
dkh_p.out:    Total Energy =                         -2.8679760471040372
nrl_0.out:    Total Energy =                         -2.8612063084748343
nrl_p.out:    Total Energy =                         -2.8678772073375063
x2c_0.out:    Total Energy =                         -2.8613053511779820
x2c_p.out:    Total Energy =                         -2.8613053511779745
```

The suffix `_0` and `_p` stand for "no field" and "dipolar field" respectively.

**Comment by loriab** [comment]
I'd hazard we need something like 
```        } else if (options_.get_str("RELATIVISTIC") == "X2C") {
            outfile->Printf(" OEINTS: Using relativistic (X2C) overlap, kinetic, and potential integrals.\n");

            if (!rel_basisset_) {
                throw PSIEXCEPTION("OEINTS: X2C requested, but relativistic basis was not set.");
            }    
            X2CInt x2cint;
            SharedMatrix so_overlap_x2c = so_overlap();
            SharedMatrix so_kinetic_x2c = so_kinetic();
            SharedMatrix so_potential_x2c = so_potential();
            x2cint.compute(basisset_, rel_basisset_, so_overlap_x2c, so_kinetic_x2c, so_potential_x2c);

            potential_mat->add(so_potential_x2c);
        }    
```
about [here](https://github.com/psi4/psi4/blob/master/psi4/src/psi4/libmints/mintshelper.cc#L1338) but that's not a literal fix. @fevangelista, is that on the right track?

**Comment by susilehtola** [comment]
Relativity does add some further complications into the mix. IIRC the X2C transform would have to be computed differently. (Whether it's an electric or magnetic field depends on the frame!)

**Comment by fevangelista** [comment]
Thank you for bringing this up. External fields are not currently implemented and I think psi4 should print out a warning if both x2c and perturb_h are invoked in the input. I am not 100% sure about how to implement perturbations but I suspect the right way to do it is to add the external potential to the nuclear potential before the X2C procedure. This way when one solves for the Dirac equation the external potential is included from the beginning. Let me see what I can find out in the literature...

**Comment by fevangelista** [comment]
@loriab: after looking at the code you linked to above, I wonder if the DKH is also implemented correctly. It seems to me like we add the perturbation after determining the DKH potential.

**Comment by tovrstra** [comment]
Thanks for the insightful comments. @fevangelista @loriab @susilehtola Would QM/MM point charges entering the Hamiltonian or other perturbations have similar issues? Also, when computing derivatives of the energy, e.g. forces on nuclei, these will also be affected by the relativistic corrections.

**Comment by robertodr** [comment]
The picture change transformation error should only be relevant in the nuclear region. If you have point charges in a QM/MM embedding fashion, these should be so far away from the QM regions that picture-change transforming that potential would have a negligible numerical effect. This is the approached  used [here](https://pubs.acs.org/doi/10.1021/acs.jctc.7b00162) As far as I know, no one ever tested this assumption. The authors of this paper note:

> A note should be added regarding the use of X2C Hamiltonians. In this case, the one-electron part of the Fock matrix is transformed (see, e.g., ref 7 for details). Strictly, all external fields (including a field from a environment) should also be transformed. Currently, our PE-X2C-DFT implementation involves transformation only of the vacuum terms. Thus, the PE operator is left untransformed, and the untransformed operator is carried over to the linear response formalism described below (following the general implementation in the DIRAC program, all property operators are always transformed).

**Comment by susilehtola** [comment]
I guess the question is: is the electric field operator transformed in Psi4?

**Comment by fevangelista** [comment]
I think I got it in the new PR I am working on right now (#1770). Here are the X2C energies w and w/o the field. I think the difference looks similar to DKH

```
x2c_0.out:    Total Energy =                         -2.8613053511779594
x2c_p.out:    Total Energy =                         -2.8679759489083452
```

**Comment by tovrstra** [comment]
Thanks! That is good news. A student ran the same test with DIRAC and got the following outputs:

```
x2c_0.out              -2.8612839050684364
x2c_p.out (field=0.1)  -2.8679305399790476
```

SCF convergence and other settings might differ a little.

**Comment by loriab** [comment]
fixed by #1770
